src/pubsub/ua_pubsub_reader.c

```diff
@@ -773,69 +774,78 @@ UA_DataSetReader *UA_ReaderGroup_findDSRbyId(UA_Server *server, UA_NodeId identi
 void UA_ReaderGroup_subscribeCallback(UA_Server *server, UA_ReaderGroup *readerGroup) {
     UA_PubSubConnection *connection =
         UA_PubSubConnection_findConnectionbyId(server, readerGroup->linkedConnection);
-    UA_ByteString buffer;
-    if(UA_ByteString_allocBuffer(&buffer, 512) != UA_STATUSCODE_GOOD) {
-        UA_LOG_INFO(&server->config.logger, UA_LOGCATEGORY_SERVER, "Message buffer alloc failed!");
-        return;
+    UA_ByteString buffer[UA_MAX_RECEIVE_MESSAGES];
+
+    for(UA_UInt32 i = 0; i < UA_MAX_RECEIVE_MESSAGES; i++) {
+        if(UA_ByteString_allocBuffer(&buffer[i], 512) != UA_STATUSCODE_GOOD) {
+            UA_LOG_INFO(&server->config.logger, UA_LOGCATEGORY_SERVER, "Message buffer alloc failed!");
+            return;
+        }
     }
 
-    connection->channel->receive(connection->channel, &buffer, NULL, 1000);
-    if(buffer.length > 0) {
-        if (readerGroup->config.rtLevel == UA_PUBSUB_RT_FIXED_SIZE) {
-            /* Considering max DSM as 1
-             * TODO:
-             * Process with the static value source
-             */
-            UA_DataSetReader *dataSetReader = LIST_FIRST(&readerGroup->readers);
-            /* Decode only the necessary offset and update the networkMessage */
-            if(UA_NetworkMessage_updateBufferedNwMessage(&dataSetReader->bufferedMessage, &buffer) != UA_STATUSCODE_GOOD) {
-                UA_LOG_INFO(&server->config.logger, UA_LOGCATEGORY_SERVER,
-                             "PubSub receive. Unknown field type.");
-                UA_ByteString_deleteMembers(&buffer);
-                return;
-            }
+    UA_UInt16 *messagesReceived = UA_UInt16_new();
+    *messagesReceived           = UA_MAX_RECEIVE_MESSAGES;
+    connection->channel->receive(connection->channel, buffer, NULL, 1000, messagesReceived);
+    for (UA_UInt16 iterator = 0; iterator < *messagesReceived; iterator++) {
```

Comment:
For the loop you can simply check if the decode position is at the end of the received buffer.